craft_application/models/project.py

```diff
@@ -91,10 +91,21 @@ class Project(CraftBaseModel):
     source_code: AnyUrl | None
     license: str | None
 
+    adopt_info: str | None
+
     parts: dict[str, dict[str, Any]]  # parts are handled by craft-parts
 
     package_repositories: list[dict[str, Any]] | None
 
+    @pydantic.root_validator(pre=True)
+    @classmethod
+    def _validate_version(cls, values: dict[str, Any]) -> dict[str, Any]:
+        if values.get("version") is None and values.get("adopt-info") is None:
+            raise ValueError(
+                "Required field 'version' is not set and 'adopt-info' not used"
+            )
+        return values
```

Comment:
```suggestion
    def _validate_adopt_info(cls, values: dict[str, Any]) -> dict[str, Any]:
    	if values.get("adopt-info") is not None:
    		if values.get("version") is None:
            raise ValueError(
                "Required field 'version' is not set and 'adopt-info' not used"
            )
    		return values
    	if values.get("adopt-info") not in values.get("parts", {}):
    		raise ValueError("'adopt-info' does not reference a valid part.")
        return values
```

Since we're adopting a part, we should be able to fail early if the user makes `adopt-info` refer to something that's not a part.

...at least I think we should?

Reply:
Validating both cases and added a corresponding unit test.